### PR TITLE
[SLP][REVEC] Make Instruction::Select support vector instructions.

### PR DESCRIPTION
**Pull request by HanKuanChen**

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -9713,6 +9713,17 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
         }
         VecCost = std::min(VecCost, IntrinsicCost);
       }
+      if (auto *SI = dyn_cast<SelectInst>(VL0)) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:9716]: Why this is specific only for selects?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -13196,6 +13207,15 @@ Value *BoUpSLP::vectorizeTree(TreeEntry *E, bool PostponedPHIs) {
           False = Builder.CreateIntCast(False, VecTy, GetOperandSignedness(2));
       }
 
+      unsigned CondNumElements = getNumElements(Cond->getType());
+      unsigned TrueNumElements = getNumElements(True->getType());
+      if (CondNumElements != TrueNumElements) {
+        // When the return type is i1 but the source is fixed vector type, we
+        // need to duplicate the condition value.
+        Cond = Builder.CreateShuffleVector(
+            Cond, createReplicatedMask(TrueNumElements / CondNumElements,
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13216]: What if CondNumElements > TrueNumElements?

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13216]: Cannot. `CondNumElements` <= `TrueNumElements`.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13216]: Add assertion

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13219]: Add in getEntryCost, including comments.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13219]: Add here too

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13223]: Done.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13223]: Can it be that `TrueNumElements % CondNumElements != 0`? Will you handle it correctly?

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13223]: Cannot. They are the operands of Instruction::Select. The number of elements must be the same.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13223]: Add the assertions, please, in all branches

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13225]: Done

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -13196,6 +13212,21 @@ Value *BoUpSLP::vectorizeTree(TreeEntry *E, bool PostponedPHIs) {
           False = Builder.CreateIntCast(False, VecTy, GetOperandSignedness(2));
       }
 
+      unsigned CondNumElements = getNumElements(Cond->getType());
+      unsigned TrueNumElements = getNumElements(True->getType());
+      assert(TrueNumElements % CondNumElements == 0 &&
+             "Cannot vectorize Instruction::Select");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:13218]: It does not prove that TrueNumElements >= CondNumElements

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**